### PR TITLE
Presets for Betaflight Configurator

### PR DESCRIPTION
**Pull request by limonspb**
To add your presets, check [this repo] readme: (https://github.com/betaflight/firmware-presets) with presets.

Cool ideas for later improvements - next PRs:

- Dropdown list of picked presets
- Load preset from a local file and show a preset dialog with options ETC.
- Add search order (rank?) #$ attribute for presets - so that we could move, for example, defaults down in the search results

## Remaining tasks for the last stage:

- [x] Ignore `#$ HIDDEN: true` presets - don't show in search
- [x] Warning and DIsclaimer dialogs when applying presets of author used #$ WARNING, #$ DISCLAIMER or included messages options
- [x] When scrolling presets only search parameters should stick to the top, not the whole top panel
- [x] Proper sizing of the presets tiles without unused space on the right
- [x] Allow different sources/repositories for presets - user dialog that allows using new sources of presets besides official Betaflight.
- [x] Official Betaflight presets source should be pointing at api.betaflight.com
- [x] Make popup windows with the preset wider
- [x] Show CLI button that runs the `#include` and `#regions` algorithm and shows the CLI commands that are going to be applied
- [x] Add `#discussion` link for the preset popup window
- [x] Small warning message at the top of the presets tab
- [x] Button for presets Wiki at the top
- [x] Add Backup load/save buttons to save/load the whole users dump into a file.
- [x] Rename buttons at the bottom and reorder them
- [x] Move presets Tab higher
- [x] Lazy CLI initialization - enter CLI mode only when the user actually applies preset
- [x] CLI errors catch - when preset is being applied with errors - make a warning.
- [x] Scroll the page down when CLI is shown.
- [x] Rename Regions -> Options in UI and in the code. 
- [x]  Make search line searching words, not the whole phrase
- [x] TUNE should be selected by default as well as the correct FW version for search.
- [x] Warning if applying preset from a wrong FW version
- [x] CLI should be available for copy text as well as CLI in the detailed dialog preview
- [x] Add magnifying glass icon next to the search string
- [x] When no presets are found show little element "No presets found for selected search parameters"
- [x] When too many presets found - show little element at the end "60 showed presets, more was found"
- [x] Animation on applying a preset
- [x] Disable save/cancel buttons when no presets are picked
- [x] Check Android
- [x] Code cleanup

### CODE REVIEW

File: src/js/serial_backend.js
Comment (haslinghuis) [inline src/js/serial_backend.js:701]:
```suggestion
    if (GUI.active_tab !== 'cli' && GUI.active_tab !== 'presets') {
```

File: src/tabs/presets/CliEngine.js
Comment (haslinghuis) [inline src/tabs/presets/CliEngine.js:106]: Please remove comment

File: src/tabs/presets/CliEngine.js
Comment (haslinghuis) [inline src/tabs/presets/CliEngine.js:155]: This is also used in cli, can we make this into util module function?

File: src/tabs/presets/PresetTitlePanel.js
Comment (haslinghuis) [inline src/tabs/presets/PresetTitlePanel.js:24]: Single statement can be a oneliner:
```suggestion
        this._domWrapperDiv.on("click", () => presetsDetailedDialog.open(this._preset));
```

File: src/tabs/presets/PresetsDetailedDialog.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsDetailedDialog.js:72]: Could be a oneliner:
```suggestion
        this._domButtonApply.on("click", () => this._onApplyButtonClicked());
```

File: src/tabs/presets/PresetsDetailedDialog.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsDetailedDialog.js:76]: Could be a oneliner:
```suggestion
        this._domButtonCancel.on("click", () => this._onCancelButtonClicked());
```

File: src/tabs/presets/PresetsDetailedDialog.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsDetailedDialog.js:88]: same

File: src/tabs/presets/PresetsDetailedDialog.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsDetailedDialog.js:91]: same

File: src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js:21]: same

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Maybe it's just personal preference. Ignore if you like, I just like the density make the code more clear to oversee. Good work!
Did a quick review.

File: src/tabs/presets/CliEngine.js
Comment (limonspb) [reply src/tabs/presets/CliEngine.js:155]: The whole CLI engine class theoretically should be reused for CLI tab... However CLI tab code is a huuge mess at least to my small brain. It took me good 5 days to separate the minimum functionality of CLI to here.

I hope someone could help to merge CLI with CLI engine and make necessary modifications for CLI Engine for that. Or it can be in the future improvements...

Comment (haslinghuis) [reply src/tabs/presets/CliEngine.js:155]: Guess it's best to make a util file first, and do the merging of cli in another PR as this is rather a critical part of configurator and keep this PR clean.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> PASS
 - PR is not in draft -> FAIL
 - Assigned to a milestone -> FAIL
 - cooling off period lapsed (9/06/2021 05:32:49) -> PASS
 - commit count less or equal to three (1 commits) -> PASS
 - 'Don't merge' label NOT found -> PASS
 - 'Needs coordination with betaflight' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver (None) -> FAIL
 - approver count at least three (0 approvers) -> FAIL

**Comment by limonspb** [comment]
The work is mainly done and should be working. Need a couple of touches here and there plus a code beautifying. I will convert draft to a pull request when ready for review.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - PR is not in draft -> FAIL
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
@limon please rebase your local branch. `stm32.js` is already updated and should not giving this errors in here anyway.

**Comment by limonspb** [comment]
Not ready for review yet. Just building for people who want to try Configurator with presets

**Review by haslinghuis — DISMISSED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/js/VirtualFC.js
Comment (haslinghuis) [inline src/js/VirtualFC.js:10]: This change has nothing to do with presets?

**Review by haslinghuis — CHANGES_REQUESTED** [review]

File: src/js/VirtualFC.js
Comment (limonspb) [reply src/js/VirtualFC.js:10]: Yeah I guess so. Should I remove,l? Figured we should change it to 4.3 anyways.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by haslinghuis — APPROVED** [review]

**Comment by ctzsnooze** [comment]
Ready to merge and tested.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by Asizon — APPROVED** [review]
I have no words, really awesome and profesional work, thankyou @limonspb

**Review by blckmn — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (PASS)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Pull request merged**
